File: src/Zahy.Settlement/Zahy.Settlement.Domain/BankAccount.cs
using System;
using Volo.Abp;
using Volo.Abp.Domain.Entities.Auditing;

namespace Zahy.Settlement;

/// <summary>
/// A bank account in the accountant-managed registry (Track B). Master data the accountant adds via
/// settings; each bank is bound to a ledger sub-account <see cref="Code"/> (110x) beneath the 1100
/// parent. Payments may route to a bank → the receipt debits this bank's sub-account instead of the
/// generic 1100. The full <see cref="AccountNumber"/> is stored but only ever shown masked.
///
/// COMPUTE/REGISTRY ONLY — adding a bank does NOT mutate the production chart of accounts; the live
/// chart change stays gated (<c>SettlementEngineOptions.BankRegistryLiveChartEnabled</c>, default OFF).
/// </summary>
public class BankAccount : FullAuditedAggregateRoot<Guid>
{
    /// <summary>Ledger sub-account code (1101–1149) under the 1100 parent. Assigned once, never reused while active.</summary>
    public string Code { get; private set; } = string.Empty;

    public string Name { get; private set; } = string.Empty;

    /// <summary>Full account number / IBAN — stored whole, displayed masked (last 4).</summary>
    public string AccountNumber { get; private set; } = string.Empty;

    public string Currency { get; private set; } = SettlementConsts.DefaultCurrency;

    public BankAccountStatus Status { get; private set; } = BankAccountStatus.Active;

    /// <summary>
    /// Optional gateway-route key for the FUTURE auto-route seam — a gateway payment WOULD locate its
    /// mapped bank via this value. SEAM ONLY: not consumed by any live gateway (gated phase).
    /// </summary>
    public string? GatewayMapping { get; private set; }

    protected BankAccount()
    {
    }

    public BankAccount(
        Guid id,
        string code,
        string name,
        string accountNumber,
        string currency,
        string? gatewayMapping = null)
        : base(id)
    {
        Code = Check.NotNullOrWhiteSpace(code, nameof(code), SettlementLedgerAccountConsts.MaxCodeLength).Trim();
        Name = NormalizeName(name);
        AccountNumber = NormalizeAccountNumber(accountNumber);
        Currency = Check.NotNullOrWhiteSpace(currency, nameof(currency), SettlementBankAccountConsts.MaxCurrencyLength).Trim();
        GatewayMapping = NormalizeGateway(gatewayMapping);
        Status = BankAccountStatus.Active;
    }

    public BankAccount Update(string name, string accountNumber, string currency, string? gatewayMapping)
    {
        Name = NormalizeName(name);
        AccountNumber = NormalizeAccountNumber(accountNumber);
        Currency = Check.NotNullOrWhiteSpace(currency, nameof(currency), SettlementBankAccountConsts.MaxCurrencyLength).Trim();
        GatewayMapping = NormalizeGateway(gatewayMapping);
        return this;
    }

    public BankAccount Deactivate()
    {
        Status = BankAccountStatus.Inactive;
        return this;
    }

    public BankAccount Activate()
    {
        Status = BankAccountStatus.Active;
        return this;
    }

    /// <summary>Display value — full number masked to the last 4 characters.</summary>
    public string MaskedAccountNumber => BankLedgerCoding.Mask(AccountNumber);

    private static string NormalizeName(string name)
    {
        if (string.IsNullOrWhiteSpace(name))
        {
            throw new BusinessException(SettlementBankAccountErrorCodes.EmptyName);
        }

        return Check.NotNullOrWhiteSpace(name, nameof(name), SettlementBankAccountConsts.MaxNameLength).Trim();
    }

    private static string NormalizeAccountNumber(string accountNumber)
    {
        if (string.IsNullOrWhiteSpace(accountNumber))
        {
            throw new BusinessException(SettlementBankAccountErrorCodes.EmptyAccountNumber);
        }

        return Check.NotNullOrWhiteSpace(
            accountNumber, nameof(accountNumber), SettlementBankAccountConsts.MaxAccountNumberLength).Trim();
    }

    private static string? NormalizeGateway(string? gatewayMapping)
    {
        if (string.IsNullOrWhiteSpace(gatewayMapping))
        {
            return null;
        }

        return Check.Length(
            gatewayMapping.Trim(), nameof(gatewayMapping), SettlementBankAccountConsts.MaxGatewayMappingLength);
    }
}

File: frontend/partner-portal-web/src/lib/banks/bankAccount.ts
import { ReportAccount } from "@/lib/reports/settlementReports";

/**
 * Track B — accountant-managed bank registry (mock). Banks are master data the accountant adds via
 * settings; each bank binds to a ledger sub-account code (1101, 1102, …) BENEATH the 1100 "Cash / Bank"
 * parent. This mirrors the backend `BankLedgerCoding` scheme. The frontend chart is name-based, so the
 * 1100 parent maps to `ReportAccount.Cash`; bank sub-accounts use their numeric 110x code as the id.
 *
 * MOCK/FLAGGED: nothing here mutates the production chart of accounts or the trial-balance money math —
 * the live chart change stays gated (CTO/accountant sign-off, same as 1250 / backend flag).
 */

export type BankAccountStatus = "Active" | "Inactive";

export interface BankAccount {
  id: string;
  /** Ledger sub-account code (1101–1149) under the 1100 parent. */
  code: string;
  name: string;
  /** Full account number / IBAN — stored whole, shown MASKED (see maskAccountNumber). */
  accountNumber: string;
  currency: string;
  status: BankAccountStatus;
  /** Optional future gateway-route key — SEAM ONLY, not wired to any live gateway. */
  gatewayMapping?: string;
}

export interface BankAccountInput {
  name: string;
  accountNumber: string;
  currency: string;
  gatewayMapping?: string;
}

/** The 1100 parent (name-based) that bank sub-accounts roll up to in the frontend chart. */
export const BANK_PARENT_ACCOUNT = ReportAccount.Cash;

/** The numeric parent code shown in the registry UI ("Bank / Cash Clearing"). */
export const BANK_PARENT_CODE = "1100";

export const BANK_SUBCODE_FIRST = 1101;
export const BANK_SUBCODE_LAST = 1149;

/** True for a reserved bank sub-account code (1101–1149) beneath the 1100 parent. */
export function isBankSubAccount(code: string | undefined | null): boolean {
  if (!code || code.length !== 4) return false;
  const n = Number(code);
  return Number.isInteger(n) && n >= BANK_SUBCODE_FIRST && n <= BANK_SUBCODE_LAST;
}

/** Next free sub-code (1101, 1102, …) given the codes in use; reuses the lowest gap first. */
export function nextBankCode(existingCodes: readonly string[]): string {
  const used = new Set<number>();
  for (const c of existingCodes) {
    const n = Number(c);
    if (Number.isInteger(n)) used.add(n);
  }
  for (let n = BANK_SUBCODE_FIRST; n <= BANK_SUBCODE_LAST; n++) {
    if (!used.has(n)) return String(n);
  }
  throw new Error("Bank sub-account range (1101–1149) exhausted.");
}

/** Mask an account number / IBAN for display — keep the last 4 characters, mask the rest. */
export function maskAccountNumber(accountNumber: string | undefined | null): string {
  const value = (accountNumber ?? "").trim();
  if (value.length <= 4) return value;
  return "•".repeat(value.length - 4) + value.slice(-4);
}

/**
 * GATEWAY AUTO-ROUTE — SEAM ONLY (not wired). A future gateway payment WOULD auto-locate its mapped
 * bank via `gatewayMapping` and auto-reconcile. This pure lookup proves the seam is reachable; there is
 * NO real gateway integration here (that is the separately-gated gateway phase).
 */
export const GATEWAY_AUTO_ROUTE_WIRED = false;

export function resolveGatewayBankCode(
  gatewayKey: string | undefined | null,
  banks: readonly BankAccount[],
): string | null {
  const key = (gatewayKey ?? "").trim();
  if (!key) return null;
  const match = banks.find(
    (b) => (b.gatewayMapping ?? "").trim().toLowerCase() === key.toLowerCase(),
  );
  return match ? match.code : null;
}
